### PR TITLE
nixos/openvpn: user, forwarding and local state dir

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

We change a few things here:

a) Create the local openvpn user and group so that openvpn can drop privileges.
This is not switched on by default.

b) Define an option to enable IP forwarding which would have to be defined
outside of the openvpn module.

c) Create a local state and runtime directories for anything openvpn might create (replay
persistence logs, current connections, etc).

d) Launch openvpn in a separate ```openvpn.slice``` slice and activate from a dedicated ```openvpn.target``` target. The former allows one to easily check the logs across all instances, apply resource limitations and stop all instances. The latter allows you to easily start all the instances.

NOTE: I *strongly* recommend looking at this diff while ignoring whitespace due to the indent changes (options were indented further due to the introduction of a new option) - https://github.com/NixOS/nixpkgs/pull/65878/files?w=1

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @viric

### CODE REVIEW

**Review by teto — APPROVED** [review]
looks good to me but I don't use openvpn.
can it work without `net.ipv4.ip_forward`  ? or is like for firewalls and default to the secure version. 
what about ipv6 ?

**Comment by peterhoeg** [comment]
>   can it work without net.ipv4.ip_forward ? or is like for firewalls and default to the secure version.

Sure it works, but without it you can only reach the machine running OpenVPN and nothing on the network behind it so both cases are equally valid.

>   what about ipv6 ?

Yeah, that's missing - nice catch. I have changed the PR but haven't tried it out yet.

**Comment by teto** [comment]
that sounds similar to the openFirewall option: you very likely need it but we cna't enable it by default. Aren't there some UDP ports that should be open as well ? would it make sense to make ports configurable (I believe that's common to avoid scans on specific ports) ?

**Comment by peterhoeg** [comment]
>   that sounds similar to the openFirewall option: you very likely need it
>   but we cna't enable it by default.

I disagree. For random service foo, with openFirewall you *have* to either use that option *or* open the firewall yourself.

Without this option, things still work - you are probably likely to want to enable it but you don't have to and you don't have to do anything else.

>   Aren't there some UDP ports that should
>   be open as well ? would it make sense to make ports configurable (I
>   believe that's common to avoid scans on specific ports) ?

OpenVPN can run over TCP and UDP, but yes, we do need to open ports.

Currently, we expose very few of the options and basically have the user throw in a bunch of text that will end up in the configuration file. Whether that's by design or because the initial author just wanted to get up and running quickly I cannot say.

Now, port and protocol would make sense to promote to config options as it would allow us to open the firewalls through a toggle as you mention but I don't think that's part of the scope here. We definitely should do it (and add a test!) though.

**Comment by peterhoeg** [comment]
>   Now, port and protocol would make sense to promote to config options as it
>   would allow us to open the firewalls through a toggle as you mention but I
>   don't think that's part of the scope here. We definitely should do it (and
>   add a test!) though.

Actually, just having thought of it a bit more, it's going to be tricky with the port/protocol.

We can run in either server or client mode and the proto/port is specified differently depending on the mode so we need to add a mode as well and act differently based on that. Considering the flexibility/complexity in the configuration file, it might not be worth trying to model that using module options.

File: nixos/modules/services/networking/openvpn.nix
Comment (teto) [inline nixos/modules/services/networking/openvpn.nix]: optionalAttrs ?

**Comment by teto** [comment]
I just wanted to mention that since it's options I would typically expect in network-related modules. It's completely fine if it's too hard to implement or too convoluted, you know better :p

File: nixos/modules/services/networking/openvpn.nix
Comment (peterhoeg) [reply nixos/modules/services/networking/openvpn.nix]: Indeed, thanks

**Comment by peterhoeg** [comment]
> It's completely fine if it's too hard to implement or too convoluted, you know better :p

I fully with you - it makes sense to have this type of stuff standard across the board but here I think it would just add complexity for not a lot of gain.

**Comment by ivan** [comment]
For the author, reviewers, and committers: this PR was scanned and appears to add a use of the deprecated `types.string`, which emits a warning as of #66346. Before merging, please change this to another type, possibly:

- `types.str` for a single string where merging does not make sense, or cannot work
- `types.lines` for multi-line configuration or scripts where merging is possible
- `types.listOf types.str` for a mergeable list of strings

**Comment by peterhoeg** [comment]
Been running with this for ages. Should be good to go - just cleaned up the commit.

**Comment by SuperSandro2000** [comment]
Closing as stale, please re-open if you want to continue it

**Pull request closed without merging**
